File: sim2.m
p = (0.7/2.6)*0.2; %Probabilidad binomial de decaer
q = 1-p;
N0=100000000; %Número de núcleos en t=0
dt= 0.2;  %delta t escogido
gamma = 0.7/2.6; %Constante de decamiento del Na-22

grid = (5:dt:15)'; % grilla de tiempos discretizados
bingo=1;

%%%%%Teorico ley de decaimiento

Nt = zeros(length(grid),1);
for i=0:length(grid)-1
    Nt(i+1) = N0*exp(-(0.7/2.6)*i*0.2);
end

%%%%%%Exponencial
fdp = (1:length(grid))';

for i=1:length(grid)
    fdp(i)= -exp(-gamma*(grid(i)+dt)) + exp(-gamma*grid(i));
end

Facum = zeros(length(grid),1);
for i=1:length(grid)-1
    Facum(i+1)= Facum(i) + fdp(i+1);
end
%bingo = 1;
frec = zeros(length(grid),1);
for i=1:N0    
    r = random('uniform',0,1);
    for j=2:length(grid)
        if Facum(j-1) < r && r < Facum(j)
            bingo = j;
            break
        end
    end
    frec(bingo) = frec(bingo) + 1;
end


  
 
%%%%%%Binomial
N=50; %Número de intentos para la acumulada
fdpx1=zeros(N,1);
fdpx0=zeros(N,1); 



for i=1:N
    fdpx1(i) = (factorial(i)/factorial(i-1))*p*q^(i-1); %fdp binomial con x=1 y N=50
    fdpx0(i) = (factorial(i)/factorial(i))*q^(i);       %fdp binomial con x=0 y N=50
end

for i=1:N
    if fdpx1(i) > fdpx0(i)
        Fx(i) = 1;            %Para N intentos, si la fdp(x=1) es mayor que la fdp(x=0), decae
    end
end

%plot(1:N,Fx)
%hold on
%xlabel('Intentos N')
%ylabel('F(x)')
%hold off
aux=N0;
frecuencias = zeros(length(grid),1);
for i=1:length(grid)    
    
    for j=1:aux
        r = random('uniform',0,1);
        if r < q
            frecuencias(i) = frecuencias(i)+1; %Simulación de núcleos que no decaen con la binomial
        end
        
    end
    aux = frecuencias(i);
    
end



M = max(frec);

stairs(grid,frecuencias*M/N0,'b')
hold on
stairs(grid,Nt*M/N0,'r')
title('Con No=10^8')
xlabel('t(años)')
ylabel('N(t)')
stairs(grid,frec,'k')
%axis([0 5 0 M])
legend('Simulación binomial','Ley de decaimiento', 'Simulación exponencial')
hold off

File: sim2.cpp
#include <iostream>
#include <cmath>
#include <random>
#include <vector>
#include <algorithm>

void leyexpo(int N0, double gamma, double tf, double dt, std::vector<double> &Nt);
void grilla(double tf, double dt, std::vector<double> &grid);
void simexpo(double tf, double gamma, std::vector<double> &grid, double dt, std::vector<double> &frecexpo);
void simbin(int N0, double tf, std::vector<double> &frecbin);

const int N0 = 100000000;
const double dt = 0.2;
const double Gamma = 0.7/2.6;
  
int main(void)
{
  double tf = 37/dt;
  std::vector<double> grid(tf+1);
  std::vector<double> Nt(tf+1);
  std::vector<double> frecexpo(tf+1);
  std::vector<double> frecbin(tf+1);

  grilla(tf, dt, grid);
  leyexpo(N0, Gamma, tf, dt, Nt);
  simexpo(tf, Gamma, grid, dt, frecexpo);
  simbin(N0, tf, frecbin);

  double M = *std::max_element(frecexpo.begin(), frecexpo.end());
  for(int kk = 0; kk < tf; ++kk){
    std::cout << grid[kk] << "\t"
	      << Nt[kk]*M/N0 << "\t"
	      << frecexpo[kk] << "\t"
	      << frecbin[kk]*M/N0 << "\n";
      //<< frecexpo[kk] << "\t"
      //      << M << "\n";
  }
  return 0;
}

void leyexpo(int N0, double gamma, double tf, double dt, std::vector<double> &Nt)
{
  for(int ii = 0; ii < tf; ++ii){
    Nt[ii] = N0*exp(-gamma*ii*dt);
  }
}

void grilla(double tf, double dt, std::vector<double> &grid)
{ 
  for(int ii = 0; ii < tf; ++ii){
    grid[ii] = dt*ii;
  }
}

void simexpo(double tf, double gamma, std::vector<double> &grid, double dt, std::vector<double> &frecexpo)
{
  std::vector<double> fdp(tf+1);
  std::vector<double> Facum(tf+1);
  int bingo = 0;
  double r = 0.0;
  int seed = 1;
  std::mt19937 gen(seed);
  std::uniform_real_distribution<> dis(0,1); //dis(gen) para llamar el random

  for(int ii = 0; ii < tf; ++ii){
    fdp[ii] = -std::exp(-gamma*(grid[ii]+dt)) + std::exp(-gamma*grid[ii]);
  }

  for(int ii = 0; ii < tf; ++ii){
    Facum[ii+1] = Facum[ii] + fdp[ii + 1];
  }

  for(int ii = 1; ii <= N0; ++ii){
    r = dis(gen);
    for(int jj = 1; jj < tf; ++jj){
      if(Facum[jj - 1] <= r && r < Facum[jj]){
	bingo = jj;
	break;
      }
    }
    frecexpo[bingo] += 1; 
  }
}

void simbin(int N0, double tf, std::vector<double> &frecbin)
{
  double aux = N0;
  int seed = 1;
  double r = 0.0;
  double p = (0.7/2.6)*0.2;
  double q = 1-p;
  std::mt19937 gen(seed);
  std::uniform_real_distribution<> dis(0,1); //dis(gen) para llamar el random

  for(int ii = 0; ii < tf; ++ii){
    for(int jj = 1; jj <= aux; ++jj){
      r = dis(gen);
      if(r < q){
	frecbin[ii] += 1;
      }
    }
    aux = frecbin[ii];
  }
}
